File: ts/src/simulation/export.ts
/**
 * Simulation export — portable result packages (AC-452).
 *
 * Exports saved simulation results as JSON, CSV, or Markdown reports.
 * Each format includes spec, variables, results, assumptions, and warnings.
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import { loadPersistedSimulationSpec, resolveSimulationArtifact } from "./artifact-store.js";
import type { SimulationResult } from "./types.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ExportFormat = "json" | "markdown" | "csv";

export interface SimulationExportOpts {
  id: string;
  knowledgeRoot: string;
  format?: ExportFormat;
  outputDir?: string;
}

export interface SimulationExportResult {
  status: "completed" | "failed";
  format: ExportFormat;
  outputPath?: string;
  error?: string;
}

// ---------------------------------------------------------------------------
// Export
// ---------------------------------------------------------------------------

export function exportSimulation(opts: SimulationExportOpts): SimulationExportResult {
  const format = normalizeExportFormat(opts.format);
  if (!format) {
    return {
      status: "failed",
      format: "json",
      error: `Unsupported export format '${String(opts.format)}'. Use json, markdown, or csv.`,
    };
  }

  const resolved = resolveSimulationArtifact(opts.knowledgeRoot, opts.id);
  if (!resolved) {
    const simDir = join(opts.knowledgeRoot, "_simulations", opts.id);
    return { status: "failed", format, error: `Simulation '${opts.id}' not found at ${simDir}` };
  }

  const { scenarioDir, report } = resolved;

  // Load spec if available
  const specPath = join(scenarioDir, "spec.json");
  const spec = loadPersistedSimulationSpec(specPath) ?? {};

  const outputDir = opts.outputDir ?? join(scenarioDir, "exports");
  if (!existsSync(outputDir)) mkdirSync(outputDir, { recursive: true });

  switch (format) {
    case "json":
      return exportJSON(report, spec, outputDir);
    case "markdown":
      return exportMarkdown(report, spec, outputDir);
    case "csv":
      return exportCSV(report, outputDir);
  }
}

// ---------------------------------------------------------------------------
// JSON
// ---------------------------------------------------------------------------

function exportJSON(
  report: SimulationResult,
  spec: Record<string, unknown>,
  outputDir: string,
): SimulationExportResult {
  const stem = exportFileStem(report);
  const pkg = {
    id: report.id,
    name: report.name,
    family: report.family,
    description: report.description,
    spec,
    variables: report.variables ?? {},
    results: {
      score: report.summary.score,
      reasoning: report.summary.reasoning,
      dimensionScores: report.summary.dimensionScores,
      bestCase: report.summary.bestCase,
      worstCase: report.summary.worstCase,
      mostSensitiveVariables: report.summary.mostSensitiveVariables,
    },
    sweep: report.sweep ?? null,
    execution: report.execution ?? null,
    assumptions: report.assumptions ?? [],
    warnings: report.warnings ?? [],
    replayOf: report.replayOf ?? null,
    originalScore: report.originalScore ?? null,
    scoreDelta: report.scoreDelta ?? null,
    exportedAt: new Date().toISOString(),
  };

  const outputPath = join(outputDir, `${stem}_export.json`);
  writeFileSync(outputPath, JSON.stringify(pkg, null, 2), "utf-8");
  return { status: "completed", format: "json", outputPath };
}

// ---------------------------------------------------------------------------
// Markdown
// ---------------------------------------------------------------------------

function exportMarkdown(
  report: SimulationResult,
  spec: Record<string, unknown>,
  outputDir: string,
): SimulationExportResult {
  const stem = exportFileStem(report);
  const lines: string[] = [];

  lines.push(`# Simulation Report: ${report.name}`);
  lines.push("");
  lines.push(`**Family:** ${report.family}`);
  lines.push(`**Status:** ${report.status}`);
  lines.push(`**Description:** ${report.description}`);
  if (report.replayOf) {
    lines.push(`**Replay Of:** ${report.replayOf}`);
  }
  lines.push("");

  // Score
  lines.push("## Score");
  lines.push("");
  lines.push(`**Overall:** ${report.summary.score.toFixed(4)}`);
  lines.push(`**Reasoning:** ${report.summary.reasoning}`);
  lines.push("");

  // Dimension scores
  const dims = report.summary.dimensionScores ?? {};
  if (Object.keys(dims).length > 0) {
    lines.push("### Dimension Scores");
    lines.push("");
    lines.push("| Dimension | Score |");
    lines.push("|-----------|-------|");
    for (const [dim, val] of Object.entries(dims)) {
      lines.push(`| ${dim} | ${(val as number).toFixed(4)} |`);
    }
    lines.push("");
  }

  // Best/worst case
  if (report.summary.bestCase) {
    lines.push(`**Best case:** ${report.summary.bestCase.score.toFixed(4)}`);
  }
  if (report.summary.worstCase) {
    lines.push(`**Worst case:** ${report.summary.worstCase.score.toFixed(4)}`);
  }
  if (report.summary.mostSensitiveVariables?.length) {
    lines.push(`**Most sensitive:** ${report.summary.mostSensitiveVariables.join(", ")}`);
  }
  lines.push("");

  // Variables
  const vars = report.variables ?? {};
  if (Object.keys(vars).length > 0) {
    lines.push("## Variables");
    lines.push("");
    lines.push("| Variable | Value |");
    lines.push("|----------|-------|");
    for (const [key, val] of Object.entries(vars)) {
      lines.push(`| ${key} | ${JSON.stringify(val)} |`);
    }
    lines.push("");
  }

  // Sweep
  if (report.sweep) {
    lines.push("## Sweep");
    lines.push("");
    lines.push(`**Dimensions:** ${report.sweep.dimensions.length}`);
    lines.push(`**Total runs:** ${report.sweep.runs}`);
    lines.push("");
    lines.push("| Variables | Score |");
    lines.push("|-----------|-------|");
    for (const run of report.sweep.results) {
      lines.push(`| ${JSON.stringify(run.variables)} | ${run.score.toFixed(4)} |`);
    }
    lines.push("");
  }

  // Assumptions
  if (report.assumptions?.length) {
    lines.push("## Assumptions");
    lines.push("");
    for (const a of report.assumptions) lines.push(`- ${a}`);
    lines.push("");
  }

  // Warnings
  if (report.warnings?.length) {
    lines.push("## Warnings");
    lines.push("");
    for (const w of report.warnings) lines.push(`- ⚠ ${w}`);
    lines.push("");
  }

  lines.push("---");
  lines.push(`*Exported at ${new Date().toISOString()}*`);

  const outputPath = join(outputDir, `${stem}_report.md`);
  writeFileSync(outputPath, lines.join("\n"), "utf-8");
  return { status: "completed", format: "markdown", outputPath };
}

// ---------------------------------------------------------------------------
// CSV
// ---------------------------------------------------------------------------

function exportCSV(report: SimulationResult, outputDir: string): SimulationExportResult {
  const stem = exportFileStem(report);
  const dims = collectDimensionKeys(report);
  const varKeys = collectVariableKeys(report);

  // Build header
  const headers = [...varKeys, "score", ...dims.map((d) => `dim_${d}`)];

  const rows: string[][] = [];

  if (report.sweep?.results?.length) {
    // Sweep: one row per sweep run
    for (const run of report.sweep.results) {
      const row: string[] = [];
      for (const key of varKeys) {
        row.push(stringifyCsvValue(run.variables?.[key] ?? report.variables?.[key] ?? ""));
      }
      row.push(String(run.score));
      for (const dim of dims) row.push(stringifyCsvValue(run.dimensionScores?.[dim] ?? ""));
      rows.push(row);
    }
  } else {
    // Single run: one data row
    const row: string[] = [];
    for (const key of varKeys) row.push(stringifyCsvValue((report.variables ?? {})[key] ?? ""));
    row.push(String(report.summary.score));
    for (const dim of dims) row.push(stringifyCsvValue((report.summary.dimensionScores ?? {})[dim] ?? ""));
    rows.push(row);
  }

  const csv = [headers.join(","), ...rows.map((r) => r.join(","))].join("\n");
  const outputPath = join(outputDir, `${stem}_data.csv`);
  writeFileSync(outputPath, csv, "utf-8");
  return { status: "completed", format: "csv", outputPath };
}

function normalizeExportFormat(format?: ExportFormat | string): ExportFormat | null {
  if (!format) return "json";
  if (format === "json" || format === "markdown" || format === "csv") {
    return format;
  }
  return null;
}

function exportFileStem(report: SimulationResult): string {
  return report.id && report.id !== report.name ? report.id : report.name;
}

function collectDimensionKeys(report: SimulationResult): string[] {
  const keys = new Set(Object.keys(report.summary.dimensionScores ?? {}));
  for (const run of report.sweep?.results ?? []) {
    for (const key of Object.keys(run.dimensionScores ?? {})) keys.add(key);
  }
  return [...keys];
}

function collectVariableKeys(report: SimulationResult): string[] {
  const keys = new Set(Object.keys(report.variables ?? {}));
  for (const run of report.sweep?.results ?? []) {
    for (const key of Object.keys(run.variables ?? {})) keys.add(key);
  }
  return [...keys];
}

function stringifyCsvValue(value: unknown): string {
  if (value == null) return "";
  if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }
  return JSON.stringify(value);
}

File: autocontext/src/autocontext/simulation/export.py
"""Simulation export — portable result packages (AC-453, parity with TS AC-452)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from autocontext.util.json_io import read_json, write_json


def export_simulation(
    id: str,
    knowledge_root: Path,
    format: str = "json",
) -> dict[str, Any]:
    """Export a saved simulation as a portable package."""
    normalized_format = format.lower()
    if normalized_format not in {"json", "markdown", "csv"}:
        return {
            "status": "failed",
            "error": f"Unsupported export format '{format}'. Use json, markdown, or csv.",
            "format": normalized_format,
        }

    resolved = _resolve_simulation_artifact(knowledge_root, id)
    if resolved is None:
        return {"status": "failed", "error": f"Simulation '{id}' not found", "format": normalized_format}

    report, sim_dir = resolved
    spec_path = sim_dir / "spec.json"
    spec = read_json(spec_path) if spec_path.exists() else {}

    output_dir = sim_dir / "exports"
    output_dir.mkdir(parents=True, exist_ok=True)

    if normalized_format == "markdown":
        return _export_markdown(report, spec, output_dir)
    if normalized_format == "csv":
        return _export_csv(report, output_dir)
    return _export_json(report, spec, output_dir)


def _resolve_simulation_artifact(knowledge_root: Path, simulation_id: str) -> tuple[dict[str, Any], Path] | None:
    simulations_root = knowledge_root / "_simulations"
    report_path = simulations_root / simulation_id / "report.json"
    if report_path.exists():
        return read_json(report_path), report_path.parent

    if not simulations_root.exists():
        return None

    for sim_dir in simulations_root.iterdir():
        if not sim_dir.is_dir() or sim_dir.name.startswith("_"):
            continue
        replay_path = sim_dir / f"replay_{simulation_id}.json"
        if replay_path.exists():
            return read_json(replay_path), sim_dir

    return None


def _export_stem(report: dict[str, Any]) -> str:
    if report.get("replay_of"):
        return f"replay_{report.get('id', 'simulation')}"
    return str(report.get("name", "simulation"))


def _collect_dimension_keys(report: dict[str, Any]) -> list[str]:
    keys = set((report.get("summary", {}) or {}).get("dimension_scores", {}).keys())
    for row in (report.get("sweep", {}) or {}).get("results", []):
        keys.update((row.get("dimension_scores", {}) or {}).keys())
    return sorted(keys)


def _collect_variable_keys(report: dict[str, Any]) -> list[str]:
    keys = set((report.get("variables") or {}).keys())
    for row in (report.get("sweep", {}) or {}).get("results", []):
        keys.update((row.get("variables", {}) or {}).keys())
    return sorted(keys)


def _stringify_csv_value(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, str):
        return value
    return json.dumps(value, sort_keys=True)


def _export_json(report: dict[str, Any], spec: dict[str, Any], output_dir: Path) -> dict[str, Any]:
    pkg = {
        "id": report.get("id", ""),
        "name": report.get("name", ""),
        "family": report.get("family", "simulation"),
        "description": report.get("description", ""),
        "spec": spec,
        "variables": report.get("variables", {}),
        "results": report.get("summary", {}),
        "execution": report.get("execution", {}),
        "sweep": report.get("sweep"),
        "assumptions": report.get("assumptions", []),
        "warnings": report.get("warnings", []),
        "replay_of": report.get("replay_of"),
        "original_score": report.get("original_score"),
        "score_delta": report.get("score_delta"),
    }
    path = output_dir / f"{_export_stem(report)}_export.json"
    write_json(path, pkg)
    return {"status": "completed", "format": "json", "output_path": str(path)}


def _export_markdown(report: dict[str, Any], spec: dict[str, Any], output_dir: Path) -> dict[str, Any]:
    name = report.get("name", "simulation")
    lines = [
        f"# Simulation Report: {name}",
        "",
        f"**Family:** {report.get('family', 'simulation')}",
        f"**Status:** {report.get('status', 'unknown')}",
        f"**Description:** {report.get('description', '')}",
    ]
    if report.get("replay_of"):
        lines.append(f"**Replay Of:** {report.get('replay_of')}")
    lines.extend([
        "",
        "## Score",
        "",
        f"**Overall:** {report.get('summary', {}).get('score', 0):.4f}",
        f"**Reasoning:** {report.get('summary', {}).get('reasoning', '')}",
        "",
    ])

    dims = report.get("summary", {}).get("dimension_scores", {})
    if dims:
        lines.extend(["### Dimension Scores", "", "| Dimension | Score |", "|-----------|-------|"])
        for dim, val in dims.items():
            lines.append(f"| {dim} | {val:.4f} |")
        lines.append("")

    sweep_results = (report.get("sweep", {}) or {}).get("results", [])
    if sweep_results:
        lines.extend(["## Sweep Results", "", "| Variables | Score | Reasoning |", "|-----------|-------|-----------|"])
        for row in sweep_results:
            lines.append(
                f"| {json.dumps(row.get('variables', {}), sort_keys=True)} "
                f"| {row.get('score', 0):.4f} | {row.get('reasoning', '')} |"
            )
        lines.append("")

    assumptions = report.get("assumptions", [])
    if assumptions:
        lines.extend(["## Assumptions", ""])
        lines.extend(f"- {a}" for a in assumptions)
        lines.append("")

    warnings = report.get("warnings", [])
    if warnings:
        lines.extend(["## Warnings", ""])
        lines.extend(f"- ⚠ {w}" for w in warnings)
        lines.append("")

    path = output_dir / f"{_export_stem(report)}_report.md"
    path.write_text("\n".join(lines), encoding="utf-8")
    return {"status": "completed", "format": "markdown", "output_path": str(path)}


def _export_csv(report: dict[str, Any], output_dir: Path) -> dict[str, Any]:
    variable_keys = _collect_variable_keys(report)
    dimension_keys = _collect_dimension_keys(report)
    rows = []

    sweep_results = (report.get("sweep", {}) or {}).get("results", [])
    if sweep_results:
        for row in sweep_results:
            row_variables = {**(report.get("variables") or {}), **(row.get("variables") or {})}
            row_dimensions = row.get("dimension_scores", {}) or {}
            rows.append({
                **{key: _stringify_csv_value(row_variables.get(key)) for key in variable_keys},
                "score": _stringify_csv_value(row.get("score")),
                "reasoning": _stringify_csv_value(row.get("reasoning")),
                **{key: _stringify_csv_value(row_dimensions.get(key)) for key in dimension_keys},
            })
    else:
        summary = report.get("summary", {}) or {}
        rows.append({
            **{key: _stringify_csv_value((report.get("variables") or {}).get(key)) for key in variable_keys},
            "score": _stringify_csv_value(summary.get("score")),
            "reasoning": _stringify_csv_value(summary.get("reasoning")),
            **{key: _stringify_csv_value((summary.get("dimension_scores", {}) or {}).get(key)) for key in dimension_keys},
        })

    headers = [*variable_keys, "score", "reasoning", *dimension_keys]
    csv_lines = [",".join(headers)]
    for row in rows:
        csv_lines.append(",".join(_escape_csv(row.get(header, "")) for header in headers))

    path = output_dir / f"{_export_stem(report)}_data.csv"
    path.write_text("\n".join(csv_lines), encoding="utf-8")
    return {"status": "completed", "format": "csv", "output_path": str(path)}


def _escape_csv(value: Any) -> str:
    text = str(value)
    if any(char in text for char in [",", "\"", "\n"]):
        return "\"" + text.replace("\"", "\"\"") + "\""
    return text
